### PR TITLE
Remove asserts and add BuildConfig

### PR DESCRIPTION
**Pull request by dnfield**
This shouldn't be landed until https://github.com/flutter/engine/pull/8820 (it contains the same changes).

This removes any `assert` from Java code and guards an AssertionError instead with `BuildConfig.DEBUG`.

Also adds logic to GN to generate the BuildConfig class.

Fixes https://github.com/flutter/flutter/issues/31928

Addresses part of https://github.com/flutter/flutter/issues/28848

### CODE REVIEW

**Comment by dnfield** [comment]
Meant to open this as a draft - oh well.

I will be rebasing this against https://github.com/flutter/engine/pull/8820  once it lands.

**Comment by dnfield** [comment]
This is ready for review.

File: shell/platform/android/io/flutter/plugin/common/StandardMessageCodec.java [hunk 1]
```diff
@@ -4,6 +4,10 @@
 
 package io.flutter.plugin.common;
 
+import android.util.Log;
```
Comment (mklim) [inline shell/platform/android/io/flutter/plugin/common/StandardMessageCodec.java]: nit: Extra import?

File: tools/gen_android_buildconfig.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+
+// THIS FILE IS AUTO_GENERATED
+// DO NOT EDIT THE VALUES HERE - SEE $flutter_root/tools/gen_android_buildconfig.py
+package io.flutter;
+
+public final class BuildConfig {{
+  public final static boolean DEBUG = {0};
+  public final static boolean PROFILE = {1};
+  public final static boolean RELEASE = {2};
```
Comment (mklim) [inline tools/gen_android_buildconfig.py]: nit: Would also be a bit safer to add a private constructor here, just to make sure this is only used for the constants.

Comment (dnfield) [reply tools/gen_android_buildconfig.py]: Done

File: shell/platform/android/io/flutter/plugin/common/StandardMessageCodec.java (hunk 1 above)
Comment (dnfield) [reply shell/platform/android/io/flutter/plugin/common/StandardMessageCodec.java]: Done

File: shell/platform/android/BUILD.gn [hunk 3]
```diff
@@ -89,6 +89,26 @@ shared_library("flutter_shell_native") {
...
+
+  args = [
+    "--out",
+    rebase_path(build_config_java),
+    "--runtime-mode",
+    flutter_runtime_mode,
+  ]
+
+  build_config = []
+  assert(build_config == [] || true)
```
Comment (chinmaygarde) [inline shell/platform/android/BUILD.gn]: I dont think you meant to leave this in here.

Comment (dnfield) [reply shell/platform/android/BUILD.gn]: You are correct.

**Comment by dnfield** [comment]
@mehmetf this is going to need something for the google3 build.  Maybe we can talk tomorrow about how to make sure this works

**Pull request merged**